File: benchmarks/ocaml/qsort.ml
type elem = int

let n = 32
let mask = 0xFFFF_FFFF

let bad_rand seed =
  Int.logand ((seed * 1664525) + 1013904223) mask
;;

let mk_random_array seed n =
  let xs = Array.make n 0 in
  let cur = ref seed in
  for i = 0 to n - 1 do
    xs.(i) <- !cur;
    cur := bad_rand !cur
  done;
  xs
;;

let check_sorted_checksum xs =
  let len = Array.length xs in
  if len = 0 then 0
  else begin
    for i = 0 to len - 2 do
      if xs.(i) > xs.(i + 1) then failwith "array is not sorted"
    done;
    xs.(len - 1) + len
  end
;;

let swap xs i j =
  let tmp = xs.(i) in
  xs.(i) <- xs.(j);
  xs.(j) <- tmp
;;

let partition xs lo hi =
  let mid = (lo + hi) / 2 in
  if xs.(mid) < xs.(lo) then swap xs lo mid;
  if xs.(hi) < xs.(lo) then swap xs lo hi;
  if xs.(mid) < xs.(hi) then swap xs mid hi;
  let pivot = xs.(hi) in
  let i = ref lo in
  for j = lo to hi - 1 do
    if xs.(j) < pivot then begin
      swap xs !i j;
      incr i
    end
  done;
  swap xs !i hi;
  !i
;;

let rec qsort_aux xs lo hi =
  if lo < hi then begin
    let mid = partition xs lo hi in
    qsort_aux xs lo (mid - 1);
    qsort_aux xs (mid + 1) hi
  end
;;

let qsort xs =
  let len = Array.length xs in
  if len > 0 then qsort_aux xs 0 (len - 1)
;;

let sort_and_checksum i =
  let xs = mk_random_array i i in
  qsort xs;
  check_sorted_checksum xs
;;

let bench () =
  let acc = ref 0 in
  for _ = 0 to n - 1 do
    for i = 0 to n - 1 do
      acc := !acc + sort_and_checksum i
    done
  done;
  !acc
;;

let () =
  Printf.printf "%d\n" (bench ())

File: benchmarks/rust/qsort.rs
type Elem = u32;

const N: usize = 32;

fn bad_rand(seed: Elem) -> Elem {
    seed.wrapping_mul(1_664_525).wrapping_add(1_013_904_223)
}

fn mk_random_array(seed: Elem, n: usize) -> Vec<Elem> {
    let mut out = Vec::with_capacity(n);
    let mut cur = seed;
    for _ in 0..n {
        out.push(cur);
        cur = bad_rand(cur);
    }
    out
}

fn check_sorted_checksum(xs: &[Elem]) -> u64 {
    if xs.is_empty() {
        return 0;
    }
    for i in 0..xs.len() - 1 {
        assert!(xs[i] <= xs[i + 1], "array is not sorted");
    }
    xs[xs.len() - 1] as u64 + xs.len() as u64
}

fn partition(xs: &mut [Elem], lo: usize, hi: usize) -> usize {
    let mid = (lo + hi) / 2;
    if xs[mid] < xs[lo] {
        xs.swap(lo, mid);
    }
    if xs[hi] < xs[lo] {
        xs.swap(lo, hi);
    }
    if xs[mid] < xs[hi] {
        xs.swap(mid, hi);
    }
    let pivot = xs[hi];
    let mut i = lo;
    let mut j = lo;
    while j < hi {
        if xs[j] < pivot {
            xs.swap(i, j);
            i += 1;
        }
        j += 1;
    }
    xs.swap(i, hi);
    i
}

fn qsort_aux(xs: &mut [Elem], lo: usize, hi: usize) {
    if lo < hi {
        let mid = partition(xs, lo, hi);
        if mid > 0 {
            qsort_aux(xs, lo, mid - 1);
        }
        qsort_aux(xs, mid + 1, hi);
    }
}

fn qsort(xs: &mut [Elem]) {
    if !xs.is_empty() {
        qsort_aux(xs, 0, xs.len() - 1);
    }
}

fn sort_and_checksum(i: usize) -> u64 {
    let mut xs = mk_random_array(i as Elem, i);
    qsort(&mut xs);
    check_sorted_checksum(&xs)
}

fn bench() -> u64 {
    let mut acc = 0_u64;
    for _ in 0..N {
        for i in 0..N {
            acc += sort_and_checksum(i);
        }
    }
    acc
}

fn main() {
    println!("{}", bench());
}
